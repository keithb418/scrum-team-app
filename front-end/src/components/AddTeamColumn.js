import React from "react";
import { Button } from "react-bootstrap";
import FontAwesome from "react-fontawesome";

const AddTeamColumn = () => {
  return (
    <Button className="add-team-button">
      <span className="content">
        <FontAwesome
          name="plus-circle"
        />
        Add Team Column
      </span>
    </Button>
  );
};

export default AddTeamColumn;
