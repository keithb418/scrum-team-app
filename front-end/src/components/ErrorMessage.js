import React from "react";
import { Alert } from "react-bootstrap";

const ErrorMessage = ({ message }) =>
  message
    ? <div><Alert bsStyle="warning"><strong>Error:</strong> {message}</Alert></div>
    : null;

export default ErrorMessage;