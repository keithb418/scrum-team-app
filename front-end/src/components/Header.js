import React from "react";
import { Row, Col } from "react-bootstrap";

const Header = () => {
  return (
    <header>
      <Row>
        <Col xs={12}>
          <div className="page-header">
            <h1>Scrum Team</h1>
          </div>
        </Col>
      </Row>
    </header>
  );
};

export default Header;
