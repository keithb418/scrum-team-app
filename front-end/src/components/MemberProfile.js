import React, { Component } from "react";
import { connect } from "react-redux";
import { Button, Row, Col, Panel } from "react-bootstrap";
import { fetchProfileData } from "../actions/shared";
import FontAwesome from "react-fontawesome";
import { Link } from "react-router-dom";
import PropTypes from "prop-types";
import { capitalizeFirstLetter } from "../utils/stringHelpers";
import Loading from "./Loading";
import Profile from "./Profile";
import Skills from "./Skills";

export class MemberProfile extends Component {
  constructor () {
    super();
  }
 
  componentDidMount () {
    this.props.fetchProfileData(this.props.match.params.id);
  }
 
  render () {
    const { teamMember, team } = this.props;
    return !teamMember ?
        <Loading /> :
        <Row>
          <Col className="profile-section" md={6} mdOffset={3} xs={12}>
            <Panel.Heading>
              <h3 className="panel-title">{team.name}</h3>
            </Panel.Heading>
            <hr />
            <div className="member-profile-subheading">
              <FontAwesome name="user-circle" className="member-img" />
              <p>{teamMember.name}</p>
            </div>
            <hr />
            <Row>
              <Profile 
                member={teamMember}/>
              <Skills 
                member={teamMember}/>
            </Row>
            <hr />
            <Link to={`/member/${teamMember._id}/edit`}>
              <Button bsStyle="primary">Edit</Button>
            </Link>
            <Link to="/">
              <Button>Back to Dashboard</Button>
            </Link>
          </Col>
        </Row>;
  }
};

MemberProfile.defaultProps = {
  team: {
    name: ""
  },
  isFetching: true
};

MemberProfile.propTypes = {
  teamMember: PropTypes.object,
  team: PropTypes.object
};

const mapStateToProps = ({ teamMembers: { teamMember }, teams: { teams } }) => {
  if(teamMember) {
    const team = teams.find(team => {
      return team._id === teamMember.team;
    });
    return {
      team,
      teamMember
    };
  } else {
    return {
      teamMember
    };
  }  
};

const mapDispatchToProps = (dispatch) => ({
  fetchProfileData: (id) => dispatch(fetchProfileData(id))
});


export default connect(mapStateToProps, mapDispatchToProps)(MemberProfile);