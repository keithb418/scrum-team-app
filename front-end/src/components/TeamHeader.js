import React, { Component } from "react";
import { Button, Row, Col } from "react-bootstrap";
import { connect } from "react-redux";
import FontAwesome from "react-fontawesome";
import PropTypes from "prop-types";
import { Link } from "react-router-dom";
import { truncateString } from "../utils/stringHelpers";
import { handleDeleteTeam } from "../actions/teams";

export class TeamHeader extends Component {
  constructor () {
    super()
  }

  handleDelete = () => {
    this.props.deleteTeam(this.props.id);
  }

  render () {
    const { teamName, id, teamMembers } = this.props;
    const isDisabled = teamMembers.length > 0;
    return (
      <div className="team-header panel-heading">
        <Link to={`team/${id}/member/add`}>
          <Button className="add-team-member-btn">
            <FontAwesome name="user-plus" />
          </Button>
        </Link>
          <h2 className="col-xs-10 panel-title">{truncateString(teamName, 16)}</h2>
        <Link to={`team/${id}/edit`}>
          <Button>
            <FontAwesome name="edit" />
          </Button>
        </Link>
        <Button
          disabled={isDisabled}
          onClick={this.handleDelete}>
          <FontAwesome name="trash" />
        </Button>
      </div>
    );
  }
};

const mapDispatchToProps = dispatch => ({
  deleteTeam: (id) => dispatch(handleDeleteTeam(id))
})

TeamHeader.propTypes = {
  teamName: PropTypes.string,
  id: PropTypes.string
};

export default connect(undefined, mapDispatchToProps)(TeamHeader);