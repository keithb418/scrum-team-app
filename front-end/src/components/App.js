import React from "react";
import TeamColumnsContainer from "../containers/TeamColumnsContainer";
import AddTeamMember from "./forms/AddTeamMember";
import Header from "./Header";

const App = () => (
  <div>
    <Header />
    <TeamColumnsContainer />
    <AddTeamMember />
  </div>
);

export default App;
