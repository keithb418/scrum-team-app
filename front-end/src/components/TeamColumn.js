import React from "react";
import { connect } from "react-redux";
import TeamMember from "./TeamMember";
import TeamHeader from "./TeamHeader";

let TeamColumn = ({ id = "", teamName = "" , project = "", teamMembers = [], dispatch }) => {

  teamMembers.sort((a, b) => {
    if (b.teamLead) {
      return 1;
    }

    if (a.teamLead) {
      return -1;
    }

    return 0;
  });

  const teamMemberComponents = teamMembers.map((teamMember) =>
    <TeamMember 
      key={teamMember._id}
      id={teamMember._id} 
      name={teamMember.name} 
      teamLead={teamMember.teamLead} 
      role={teamMember.role} />
  );

  let allowDrop = (e) => {
    e.preventDefault();
  };

  let drop = (e) => {
    e.preventDefault();

    dispatch({
      type: "CHANGE_TEAM",
      teamMemberId: e.dataTransfer.getData("tmId"),
      team: id
    });
  };

  return (
    <div className="panel panel-default team-column" onDragOver={allowDrop} onDrop={drop}>
      <TeamHeader 
        id={id}
        teamName={teamName} 
        project={project} />
      <div className="panel-body team-body" onDrop={drop} onDragOver={allowDrop} >
        {teamMemberComponents}
      </div>
    </div>
  );
};

TeamColumn = connect()(TeamColumn);

export default TeamColumn;