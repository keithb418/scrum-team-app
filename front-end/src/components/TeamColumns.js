import React from "react";
import TeamColumn from "./TeamColumn";
import AddTeamColumn from "./AddTeamColumn";

const TeamColumns = ({ teams }) => {
  
  let teamColumns = teams.map((team) => 
    <TeamColumn
      key={team._id}
      id={team._id} 
      teamName={team.name} 
      project={team.project}
      teamMembers={team.teamMembers} />
  );

  return (
    <div className="team-columns">
      {teamColumns}
      <AddTeamColumn />
    </div>
  );
};

export default TeamColumns;