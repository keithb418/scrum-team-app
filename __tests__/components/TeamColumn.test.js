import React from 'react';
import { shallow } from 'enzyme';
import teamMemberData from '../fixtures/teamMembers';
import teamData from '../fixtures/teams';
import { TeamColumn } from '../../front-end/src/components/TeamColumn';

let changeTeam, allowDrop, drop, wrapper;
beforeEach(() => {
    changeTeam = jest.fn();
    allowDrop = jest.fn();
    drop = jest.fn();
    wrapper = shallow(
        <TeamColumn
            key={teamData[0]._id}
            id={teamData[0]._id}
            teamName={teamData[0].name}
            teamMembers={teamMemberData} 
            changeTeam={changeTeam} 
        />
    );
});

test('should render TeamColumn component correctly', () => {
    expect(wrapper).toMatchSnapshot();
});


test('should handle drag and drop event', () => {
    let mockEvent = {
        preventDefault: () => { },
    };
    wrapper.find('div.team-column').simulate('dragstart', mockEvent);
    wrapper.find('div.panel-body').simulate('dragover', mockEvent);
    wrapper.find('div.team-column').simulate('dragend', mockEvent);
    wrapper.find('div.team-column').simulate('drop', {
        preventDefault: () => { },
        dataTransfer : { getData: () => teamData[0].name }
    });
    allowDrop();
    drop(); 
    expect(changeTeam).toHaveBeenCalledWith(teamData[0].name, teamData[0]._id);
});
