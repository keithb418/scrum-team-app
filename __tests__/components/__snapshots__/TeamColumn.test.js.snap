// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render TeamColumn component correctly 1`] = `
<div
  className="panel panel-default team-column"
  onDragOver={[Function]}
  onDrop={[Function]}
>
  <Connect(TeamHeader)
    id="1"
    teamMembers={
      Array [
        Object {
          "_id": "1",
          "email": "mtest@mail.com",
          "experience": "4 years",
          "name": "Marc",
          "role": "Fullstack Developer",
          "skills": Array [
            "react",
          ],
          "team": "1abc",
          "teamLead": false,
        },
        Object {
          "_id": "2",
          "email": "dtest@mail.com",
          "experience": "3 years",
          "name": "Dennis",
          "role": "Fullstack Developer",
          "skills": Array [
            "react",
          ],
          "team": "2abc",
          "teamLead": false,
        },
        Object {
          "_id": "3",
          "email": "m1test@mail.com",
          "experience": "2 years",
          "name": "Matt",
          "role": "Fullstack Developer",
          "skills": Array [
            "react",
          ],
          "team": "1abc",
          "teamLead": false,
        },
      ]
    }
    teamName="Team 1"
  />
  <div
    className="panel-body team-body"
    onDragOver={[Function]}
    onDrop={[Function]}
  >
    <Connect(TeamMember)
      id="1"
      key="1"
      name="Marc"
      role="Fullstack Developer"
      teamLead={false}
    />
    <Connect(TeamMember)
      id="2"
      key="2"
      name="Dennis"
      role="Fullstack Developer"
      teamLead={false}
    />
    <Connect(TeamMember)
      id="3"
      key="3"
      name="Matt"
      role="Fullstack Developer"
      teamLead={false}
    />
  </div>
</div>
`;
